File: backend/src/main/java/com/petconnect/backend/user/application/service/AuthService.java
package com.petconnect.backend.user.application.service;

import com.petconnect.backend.exception.EntityNotFoundException;
import com.petconnect.backend.exception.InvalidPasswordResetTokenException;
import com.petconnect.backend.user.application.dto.AuthLoginRequestDto;
import com.petconnect.backend.user.application.dto.AuthResponseDto;
import com.petconnect.backend.user.application.dto.OwnerProfileDto;
import com.petconnect.backend.user.application.dto.OwnerRegistrationDto;
import com.petconnect.backend.exception.EmailAlreadyExistsException;
import com.petconnect.backend.exception.UsernameAlreadyExistsException;
import com.petconnect.backend.user.application.dto.PasswordResetRequestDto;
import com.petconnect.backend.user.application.dto.PasswordResetDto;

/**
 * Service interface for authentication and initial user registration processes.
 * Defines operations for registering new owners and authenticating existing users.
 *
 * @author ibosquet
 */
public interface AuthService {
    /**
     * Registers a new Pet Owner user in the system.
     * <p>
     * This method handles the creation of a new user entity specifically for the OWNER role.
     * It includes validation of input data (like email and username uniqueness) and
     * secure password hashing before persisting the new user.
     * </p>
     *
     * @param registrationDTO DTO containing the owner's registration details (username, email, password, phone).
     * @return A DTO representing the profile of the newly created owner upon successful registration.
     * @throws EmailAlreadyExistsException    if the provided email address is already registered in the system.
     * @throws UsernameAlreadyExistsException if the provided username is already taken.
     * @throws IllegalArgumentException       if input validation fails on the DTO.
     * @throws IllegalStateException          if required, system roles (like OWNER) are not configured correctly.
     */
    OwnerProfileDto registerOwner(OwnerRegistrationDto registrationDTO);

    /**
     * Authenticates a user based on the provided login credentials (username and password).
     * <p>
     * Verifies the credentials against stored user data. If authentication is successful,
     * it generates and returns a JSON Web Token (JWT) for later API requests.
     * </p>
     *
     * @param authLoginRequest DTO containing the user's login credentials (username and password).
     * @return An AuthResponseDto containing the username, a success message, the generated JWT,
     *         and a status flag indicating success (true).
     */
    AuthResponseDto loginUser(AuthLoginRequestDto authLoginRequest);

    /**
     * Initiates the password reset process for a user identified by their email address.
     * If the email exists, generates a unique reset token, saves it with an expiration date,
     * and triggers an email containing a reset link with the token to the user.
     * Returns void, indicating the request was processed (but not necessarily that an email was sent,
     * to avoid revealing whether an email is registered).
     *
     * @param requestDto DTO containing the user's email address.
     * @throws EntityNotFoundException (Potentially suppressed) if the email is not found, depending on security policy.
     * @throws RuntimeException if token generation or email sending fails.
     */
    void requestPasswordReset(PasswordResetRequestDto requestDto);

    /**
     * Resets the user's password using a valid, non-expired reset token.
     * Validates the token, checks password confirmation, hashes the new password,
     * updates the user's password in the database, and invalidates the token.
     *
     * @param resetDto DTO containing the reset token, new password, and confirmation.
     * @throws InvalidPasswordResetTokenException if the token is invalid, not found, or expired.
     * @throws IllegalArgumentException if the new passwords do not match or meet policy.
     * @throws EntityNotFoundException if the user associated with the token is not found (should not happen if the token is valid).
     */
    void resetPassword(PasswordResetDto resetDto);
}

File: frontend/src/services/authService.ts
import axios from 'axios';
import { API_BASE_URL } from '../config';
import {
    AuthResponse,
    ApiErrorResponse,
    UserProfile,
    OwnerRegistrationDto, 
    OwnerProfile,
} from '../types/apiTypes';

interface LoginCredentials {
  username: string; 
  password: string;
}

// Data needed for initiating password reset
interface ResetPasswordRequestData {
    email: string;
  }

  // Data needed for completing password reset
interface ResetPasswordData {
    token: string;
    newPassword: string;
    confirmPassword: string; 
}

// Simple response for password reset confirmation
interface ResetPasswordResponse {
    message: string;
}

/**
 * Attempts to log in a user by sending credentials to the backend API.
 *
 * @param {LoginCredentials} credentials - The user's login credentials (username/email and password).
 * @returns {Promise<AuthResponse>} A promise that resolves with the authentication response data (including JWT).
 * @throws {Error} Throws an error if login fails.
 */
export const loginUser = async (credentials: LoginCredentials): Promise<AuthResponse> => {
    try {
      const response = await axios.post<AuthResponse>(`${API_BASE_URL}/auth/login`, credentials, {
          headers: {
              'Content-Type': 'application/json',
          }
      });
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
          const apiError = error.response.data as ApiErrorResponse;
          console.error('API Login Error:', apiError);
          throw new Error(typeof apiError.message === 'string' ? apiError.message : apiError.error || 'Login failed');
      } else {
        console.error('Network or unexpected login error:', error);
        throw new Error('Login failed due to network or unexpected error.');
      }
    }
};

/**
 * Fetches the profile of the currently authenticated user using the provided JWT token.
 *
 * @param {string} token - The JWT token for authorization.
 * @returns {Promise<UserProfile>} A promise that resolves with the user's profile data (OwnerProfile or ClinicStaffProfile).
 * @throws {Error} Throws an error if fetching fails.
 */
export const getCurrentUserProfile = async (token: string): Promise<UserProfile> => {
    if (!token) {
        throw new Error("No authentication token provided.");
    }
    try {
        const response = await axios.get<UserProfile>(`${API_BASE_URL}/users/me`, {
            headers: {
                'Authorization': `Bearer ${token}`,
                'Content-Type': 'application/json',
            }
        });
        return response.data;
    } catch (error) {
        if (axios.isAxiosError(error) && error.response) {
            const apiError = error.response.data as ApiErrorResponse;
            console.error('API Get Profile Error:', apiError);
            throw new Error(typeof apiError.message === 'string' ? apiError.message : apiError.error || 'Failed to fetch user profile');
        } else {
            console.error('Network or unexpected get profile error:', error);
            throw new Error('Failed to fetch user profile due to network or unexpected error.');
        }
    }
};

/**
 * Registers a new Owner user by sending registration data to the backend API.
 *
 * @param {OwnerRegistrationDto} registrationData - The owner's registration details matching the backend DTO.
 * @returns {Promise<OwnerProfile>} A promise that resolves with the profile data of the newly created owner.
 * @throws {Error} Throws an error if registration fails.
 */
export const registerOwner = async (registrationData: OwnerRegistrationDto): Promise<OwnerProfile> => {
  try {
      const response = await axios.post<OwnerProfile>(`${API_BASE_URL}/auth/register`, registrationData, {
          headers: {
              'Content-Type': 'application/json',
          }
      });
      return response.data;
  } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
          const apiError = error.response.data as ApiErrorResponse;
          console.error('API Registration Error:', apiError);
           let errorMessage = 'Registration failed.';
           if (typeof apiError.message === 'string') {
               errorMessage = apiError.message;
           } else if (typeof apiError.message === 'object' && apiError.message !== null) {
               errorMessage = Object.values(apiError.message).join(' ');
           } else if (apiError.error) {
               errorMessage = apiError.error;
           }
          throw new Error(errorMessage);
      } else {
          console.error('Network or unexpected registration error:', error);
          throw new Error('Registration failed due to network or unexpected error.');
      }
  }
};

/**
 * Sends a request to the backend to initiate the password reset process for the given email.
 *
 * @param {string} email - The email address for which to request a password reset.
 * @returns {Promise<void>} A promise that resolves when the request is successfully processed by the backend.
 * @throws {Error} Throws an error if the request fails.
 */
export const requestPasswordReset = async (requestData: ResetPasswordRequestData): Promise<void> => {
    try {
        await axios.post(`${API_BASE_URL}/auth/forgot-password`, requestData, {
            headers: {
                'Content-Type': 'application/json',
            }
        });
    } catch (error) {
        if (axios.isAxiosError(error) && error.response) {
            const apiError = error.response.data as ApiErrorResponse;
            console.error('API Forgot Password Error:', apiError);
             let errorMessage = 'Failed to send password reset request.';
              if (error.response.status === 400) {
                 errorMessage = 'Invalid email format provided.';
              } else if (typeof apiError.message === 'string') {
                 errorMessage = apiError.message;
             } else if (apiError.error) {
                errorMessage = apiError.error;
             }
            throw new Error(errorMessage);
        } else {
            console.error('Network or unexpected forgot password error:', error);
            throw new Error('Password reset request failed due to network or unexpected error.');
        }
    }
};

/**
 * Resets the user's password using the provided token and new password.
 *
 * @param {ResetPasswordData} resetData - Includes token, newPassword, confirmPassword.
 * @returns {Promise<ResetPasswordResponse>} A promise resolving with a success message interface.
 * @throws {Error} Throws an error if reset fails.
 */
export const resetPassword = async (resetData: ResetPasswordData): Promise<ResetPasswordResponse> => {
    if (!resetData.token) {
        throw new Error("Reset token is missing.");
    }
    if (resetData.newPassword !== resetData.confirmPassword) {
         throw new Error("Passwords do not match.");
    }

    try {
         const response = await axios.post<ResetPasswordResponse>(`${API_BASE_URL}/auth/reset-password`, {
             token: resetData.token,
             newPassword: resetData.newPassword,
             confirmPassword: resetData.confirmPassword // Send confirmation to backend if API expects it
         }, {
             headers: {
                 'Content-Type': 'application/json',
             }
         });
         return response.data;
    } catch (error) {
        if (axios.isAxiosError(error) && error.response) {
            const apiError = error.response.data as ApiErrorResponse;
            console.error('API Reset Password Error:', apiError);
             let errorMessage = 'Failed to reset password.';
             if (error.response.status === 400 && typeof apiError.message === 'string') {
                errorMessage = apiError.message; 
             } else if (apiError.error) {
                errorMessage = apiError.error;
             }
            throw new Error(errorMessage);
        } else {
            console.error('Network or unexpected reset password error:', error);
            throw new Error('Password reset failed due to network or unexpected error.');
        }
    }
};